features/26-mfa/api.md

```diff
@@ -0,0 +1,961 @@
+
+```JSON
+{
+  "recovery_codes": ["xxx", "xxx"]
+}
+```
+
+### Description
+
+Activate the just created inactive OOB authenticator.
+
+## POST /mfa/totp/authenticate
+
+### Request body
+
+```JSON
+{
+  "otp": "123456",
+  "request_bearer_token": true,
+  "authn_session_token": "xxx"
+}
+```
+
+- `otp`: The one-time password. The OTP is tried with all active authenticators.
+- `request_bearer_token`: Whether or not to issue a bearer token
+- `authn_session_token`: The authentication session token.
+
+### Response body
+
+```JSON
+{
+  "user": { },
+  "identity": { },
+  "access_token": "xxx",
+  "mfa_bearer_token": "yyy"
+}
+```
+
+- `mfa_bearer_token`: Present only when `request_bearer_token` is `true`.
+
+## POST /mfa/oob/authenticate
+
+### Request body
+
+```JSON
+{
+  "code": "123456",
+  "request_bearer_token": true,
+  "authn_session_token": "xxx"
+}
+```
+
+- `code`: The code received from out-of-band channel. The code is tried with all valid codes.
+- `request_bearer_token`: Whether or not to issue a bearer token
+- `authn_session_token`: The authentication session token.
+
+### Response body
+
+```JSON
+{
+  "user": { },
+  "identity": { },
+  "access_token": "xxx",
+  "mfa_bearer_token": "yyy"
+}
+```
+
+- `mfa_bearer_token`: Present only when `request_bearer_token` is `true`.
+
+## POST /mfa/recovery_code/authenticate
+
+### Request body
+
+```JSON
+{
+  "code": "XXXX-YYYY-ZZZZ",
+  "authn_session_token": "xxx"
+}
+```
+
+- `code`: A recovery code.
+- `authn_session_token`: The authentication session token.
+
+### Response body
+
+```JSON
+{
+  "user": { },
+  "identity": { },
+  "access_token": "xxx"
+}
+```
+
+## POST /mfa/bearer_token/authenticate
+
+### Request body
+
+```JSON
+{
+  "bearer_token": "xxx",
+  "authn_session_token": "xxx"
+}
+```
+
+- `bearer_token`: A bearer token.
+- `authn_session_token`: The authentication session token.
+
+### Response body
+
+```JSON
+{
+  "user": { },
+  "identity": { },
+  "access_token": "xxx"
+}
+```
+
+## GET /mfa/authenticators
+
+### Request
+
+No input is required.
+
+### Response body
+
+```JSON
+{
+  "authenticators": [
+    {
+      "id": "1",
+      "type": "totp",
+      "activated_at": "2019-07-19T00:00:00.000Z",
+      "display_name": "totp-2edfdca2-3f05-41d2-bfd8-502725ee09a6"
+    },
+    {
+      "id": "2",
+      "type": "oob",
+      "activated_at": "2019-07-19T00:00:00.000Z",
+      "channel": "sms",
+      "masked_phone": "+85223******"
+    },
+    {
+      "id": "3",
+      "type": "oob",
+      "activated_at": "2019-07-19T00:00:00.000Z",
+      "channel": "email",
+      "masked_phone": "ab******@example.com"
+    }
+  ]
+}
+```
+
+## DELETE /mfa/authenticator
+
+### Request body
+
+```JSON
+{
+  "authenticator_id": "XXX"
+}
+```
+
+### Response body
+
+No response body.
+
+## POST /mfa/recovery_code/regenerate
+
+### Request body
+
+No request body
+
+### Response body
+
+```JSON
+{
+  "recovery_codes": ["xxx", "yyy"]
+}
+```
+
+## GET /mfa/recovery_code
```

Comment:
Am I right to assume to get recovery code / regenerate, a recent auth session is needed (so users will need to type password again if their sessions is not recently authed?)

How these requirements are recorded in spec for now? @louischan-oursky

Reply:
Agree we should have re-authentication for security-related operation like this.

The re-authentication is briefly mentioned [here](https://github.com/SkygearIO/features/blob/master/features/329-identity-apis/api.md#configuration). Maybe we should move the spec of re-authentication to its own document.

Should I open a feature issue for this?

Reply:
Sure Thx.

Reply:
Opened [a feature issue](https://github.com/SkygearIO/features/issues/367)